add(Search): Search for type of package

IanDuncanT opened the pull request:
Search for a type of package or all for a list. Also adds methods to print tables and get terminal size.

IanDuncanT committed:
Starting search by type

Prints out comments of package files. Still need to add seperator and parsing after that. Waiting on feedback of types and seperator.

IanDuncanT committed:
Add package search and functions for printing.

IanDuncanT committed:
Circle ci update (#25)

* Add some more tests

* Add tests

* Fix tests

* Remove bad keys

* Fix small indentation

* Update travis as well

* Update .travis.yml

Wabri commented on .travis.yml:
```diff
@@ -11,6 +11,7 @@ matrix:
       dist: trusty
-script: python yapi.py install test
+script:
+  - python yapi.py install test
```
It's probably need a script per every command of yapi to have more coverage.

IanDuncanT replied:
I will add that.

Wabri commented on modules/cross_platform.py:
```diff
@@ -1,11 +1,42 @@
+
+def term_width():
+    import os
+    rows, columns = os.popen('stty size', 'r').read().split()
```
what does this instruction do?

IanDuncanT replied:
This returns the terminal width for printing tables. The terminal width is used so that row length can be calculated and spacing used. The command 'stty size' returns 2 numbers with a space in between. ex.('24 80' for a 24x80 terminal window.) The .read().split() assigning two variables for the rows and columns length. I return the one I want for that function.

Wabri commented on yapi.py:
```diff
@@ -25,7 +26,7 @@
         language_selected = config['Languages.selected']
     except:
-        print('Config not able to be imported. Run \"python3 yapi.py config\" to fix the error')
+        print('Config not able to be imported. Run \'python3 yapi.py config\' to fix the error')
```
I need to add this into origin

Wabri commented on modules/search.py:
```diff
@@ -29,3 +29,20 @@ def search(api_url, file_extension, local_boolean, local_dir, search_pattern):
     result = list(filter(r.match, files))
     return result
+
+def searchType(os_platform, api_url, base_url, branch, file_extension, local_boolean, local_dir, package_type):
```
I dont understand what's the difference between the search functionality

IanDuncanT replied:
Search type returns all packages of a type ex. 'ide' will return Atom, VS Code, and any others that we add with that type. See the package definitions for how types are written there. I need to update the web documentation once this is merged.

IanDuncanT committed:
Add(Travis Tests):

Add travis tests for gui, full installer, download, run, search, and type search.

IanDuncanT committed:
Fix (Testing):

GUI testing always will hang in testing environments. Do not test GUI code with travis or circleci.

IanDuncanT committed:
Fix(Bash Escaping):

Add backslash to fix bash escaping when searching regex.

IanDuncanT committed:
Fix (Github API Rate Limit):

TravisCI build fails for search and type when using Github api due to rate limiting. No current solution to fix this.

IanDuncanT committed:
Removed some packages

Removed packages that are not in use

IanDuncanT committed:
Merge branch 'search-by-type-#5' of https://github.com/YetAnotherPackageInstaller/rewrite into search-by-type-#5

IanDuncanT committed:
Create (New Form):

New form is run with new before the command. python3 yapi.py new <command> <package>
This can be changed to shell files and run by changing the config section for file_extension and running the old commands

Wabri commented on modules/cross_platform.py:
```diff
@@ -1,11 +1,42 @@
     if os.path.basename(os.getcwd()) != directory:
         os.chdir(fix_path(directory, platform))
+
+def tabprint(array):
```
I think i will steal this functions and add it to origin

Wabri commented on modules/cross_platform.py:
```diff
@@ -1,11 +1,42 @@
+            result += '\n'
+    print(result)
+
+def term_width():
```
This function can be used also to avoid the overflow line when terminal is too small.

DanerSound commented on .circleci/config.yml:
```diff
@@ -6,35 +6,94 @@ workflows:
+  test-3.3-search-all:
+    <<: *37-search-all
     docker:
       - image: circleci/python:3.3-jessie
```
this first part is some kind of switch case ?

DanerSound commented on modules/cross_platform.py:
```diff
@@ -1,11 +1,42 @@
+        else:
+            pos = 0
+            result += '\n'
+    print(result)
```
I don't understand what kind of items you are appending

IanDuncanT replied:
I am printing an array of names and this takes the strings and makes a table that can be printed.

Wabri replied:
you can find the call here:
https://github.com/YetAnotherPackageInstaller/rewrite/blob/dcdd8dc818a0ff069399239a864a0aeb891cd032/yapi.py#L42
and here:
https://github.com/YetAnotherPackageInstaller/rewrite/blob/dcdd8dc818a0ff069399239a864a0aeb891cd032/yapi.py#L45
the parameter matches is a list of string so in this method the array is intended to be a list of string.
the first for is used to get the max length and add 5 (this will explain below).
the second for create a string and prevent an ugly indentation of the element in the array list.

To be more precise, from this list of string:
```
element one found at
element two not found at path
```
if you have a terminal with a width of 10 the result string will be:
```
element one
  found at
element two
  not found
  at path
```
not print more than 10 characters per line of terminal.

DanerSound commented on modules/new_installer.py:
```diff
@@ -0,0 +1,82 @@
+
+    # Downloading
+    print('--> Downloading Package File')
+    file_name = package + file_extension
```
this are support package? or the package that client wants to download

Wabri replied:
this is the name of the file plus the extension support, if the package you want to install is called yapi the full name with the extension will be yapi.sh or yapi.yp (in this new version)

IanDuncanT replied:
This is the .yp file for the package to download. It then downloads any extra files for installation (such as an archive) before running the script that doesn't need to get any resources externally.

DanerSound reviewed (commented):
all clear

Wabri commented on modules/new_installer.py:
```diff
@@ -0,0 +1,82 @@
+    print('--> Downloading Additional Resources')
+    package_file = ConfigParser(interpolation=ExtendedInterpolation())
+    package_file.read(file_name)
+    package_download = package_file['Script']['download']
```
The download section need a check because there are some packages that are not supposed to download anything, like [oh-my-zsh.sh](https://github.com/YetAnotherPackageInstaller/packages-linux-debian/blob/master/scripts/oh-my-zsh.sh).
It's not a problem to create a download field, we need to separate the download resource from the executable.
Instead of this:
```
sh -c "$(wget https://raw.githubusercontent.com/robbyrussell/oh-my-zsh/master/tools/install.sh -O -)"
```
this:
```
wget https://raw.githubusercontent.com/robbyrussell/oh-my-zsh/master/tools/install.sh -O install.sh
sh ./install.sh
```
or something like that.
We need to remember to do that although can be a problem.

IanDuncanT replied:
That's good to remember. We need to transfer the shell scripts to this new package type so that  we can run the new version.

Wabri commented on modules/new_installer.py:
```diff
@@ -0,0 +1,82 @@
+    package_file = ConfigParser(interpolation=ExtendedInterpolation())
+    package_file.read(file_name)
+    package_download = package_file['Script']['download']
+    with open(package + '.log', 'a') as package_log:
```
I have nothing against log files, but why do you think it might be useful? Maybe when users have some troubles we need to get some infos and the log file can help, am i wrong?

IanDuncanT replied:
I am outputting to log files because this allows only a few lines for output of the step and allows us to debug, but not cause clutter in the terminal for other users.

Wabri replied:
I totally agree with you

Wabri commented on scripts/test.log:
```diff
@@ -0,0 +1,10 @@
+"Download section test"
```
maybe more infos are useful for a log file, like: time, command run or something like that.

IanDuncanT replied:
This is where the output of the commands in the .yp file go. I could add some time format for it, but that will be later once I get the program working for continuous integration and scripts transferred to the new format.

IanDuncanT committed:
Fix (New Install method):

I still need to transfer scripts for new format.

IanDuncanT committed:
Fix (Byte Decode):

Use byte decoding to form string. Fix for python 3.3

IanDuncanT committed:
Fix (decode)

This doesn't work for python 3.3, but for all other versions

IanDuncanT commented:
I know this most recent will not pass search 3.3 for both individual packages and type. I need some feedback on solving this issue because I need to decode a urllib.requests object and decode it to then import it into json. I have no decoding, which works with python 3.6 and 3.7, but when I add any, it starts failing.

IanDuncanT committed:
Remove (Python 3.3):

Remove python 3.3 support as it is not officially supported. https://www.python.org/downloads/release/python-337/

IanDuncanT commented:
Closes #5 Closes #26
